src/rust/engine/fs/brfs/src/main.rs

```diff
@@ -0,0 +1,897 @@
+extern crate bazel_protos;
+extern crate clap;
+extern crate env_logger;
+extern crate errno;
+extern crate fs;
+extern crate fuse;
+extern crate futures;
+extern crate hashing;
+extern crate libc;
+#[macro_use]
+extern crate log;
+extern crate protobuf;
+extern crate time;
+
+use futures::future::Future;
+use hashing::{Digest, Fingerprint};
+use std::collections::HashMap;
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::ffi::{CString, OsStr, OsString};
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+
+const TTL: time::Timespec = time::Timespec { sec: 0, nsec: 0 };
+
+const CREATE_TIME: time::Timespec = time::Timespec { sec: 1, nsec: 0 };
+
+fn dir_attr_for(inode: Inode) -> fuse::FileAttr {
+  attr_for(inode, 0, fuse::FileType::Directory, 0x555)
+}
+
+fn attr_for(inode: Inode, size: u64, kind: fuse::FileType, perm: u16) -> fuse::FileAttr {
+  fuse::FileAttr {
+    ino: inode,
+    size: size,
+    // TODO: Find out whether blocks is actaully important
+    blocks: 0,
+    atime: CREATE_TIME,
+    mtime: CREATE_TIME,
+    ctime: CREATE_TIME,
+    crtime: CREATE_TIME,
+    kind: kind,
+    perm: perm,
+    nlink: 1,
+    uid: 0,
+    gid: 0,
+    rdev: 0,
+    flags: 0,
+  }
+}
+
+pub fn digest_from_filepath(str: &str) -> Result<Digest, String> {
+  let mut parts = str.split("-");
+  let fingerprint_str = parts
+    .next()
+    .ok_or_else(|| format!("Invalid digest: {}", str))?;
+  let fingerprint = Fingerprint::from_hex_string(fingerprint_str)?;
+  let size_bytes = parts
+    .next()
+    .ok_or_else(|| format!("Invalid digest: {}", str))?
+    .parse::<usize>()
+    .map_err(|err| format!("Invalid digest; size {} not a number: {}", str, err))?;
+  Ok(Digest(fingerprint, size_bytes))
+}
+
+type Inode = u64;
+
+const ROOT: Inode = 1;
+const DIGEST_ROOT: Inode = 2;
+const DIRECTORY_ROOT: Inode = 3;
+
+#[derive(Clone, Copy, Debug)]
+enum EntryType {
+  File,
+  Directory,
+}
+
+#[derive(Clone, Copy, Debug)]
+struct InodeDetails {
+  digest: Digest,
+  entry_type: EntryType,
+  is_executable: bool,
+}
+
+#[derive(Debug)]
+struct ReaddirEntry {
+  inode: Inode,
+  kind: fuse::FileType,
+  name: OsString,
+}
+
+enum Node {
+  Directory(bazel_protos::remote_execution::DirectoryNode),
+  File(bazel_protos::remote_execution::FileNode),
+}
+
+struct BuildResultFS {
+  store: fs::Store,
+  inode_digest_cache: HashMap<Inode, InodeDetails>,
```

Comment:
A bit about strategy here... I think that one thing that might be possible would be to have a caching wrapper around `Store`, which caches `Shared<Box<Future<T,E>>>` store results... or to use `Graph` to do that memoization for you. If you were to use a `Graph` all from within the same context, the result would be a single root with all of the filesystem requests as direct leaf children... ie, ~no cost to cycle detect.

As mentioned in slack, the advantage of caching Futures is that you can prepopulate them from a background thread which would eagerly walk the `Digest`/`Directory` to warm it. This would allow for concurrent fetching/checkouts, even when the process consuming `brfs` is not readig files concurrently.

Depending on the memoization API, the "downside" would be that all inputs would be in memory. But honestly, I expect that that would be fine for a good long while (if not forever).